app/models/concerns/entity_events.rb

```diff
@@ -0,0 +1,35 @@
+module EntityEvents
+  extend ActiveSupport::Concern
+
+  included do
+    attr_accessor :event_tags
+
+    after_create do
+      data = entity_data(attributes)
+      send_event('entity_created', data) if data.any?
+    end
+
+    after_update do
+      data = entity_data(saved_changes)
+      send_event('entity_updated', data) if data.any?
+    end
+  end
+
+  def send_event(type, data)
+    event = Events::Event.new
+      .with_type(type)
+      .with_data(default_entity_data.merge(data))
+      .with_tags(event_tags)
+
+    SendEventsToBigquery.perform_async(event.as_json)
+  end
+
+  def entity_data(changeset)
+    exportable_attrs = Rails.configuration.analytics[self.class.table_name.to_sym]
+    changeset.slice(*exportable_attrs&.map(&:to_s))
+  end
+
+  def default_entity_data
+    { table_name: self.class.table_name }
```

Comment:
I have this feeling that including the `table_name` in this manner might actually be tricky to actually decode later ... I know TVS does it this way, but I wonder if we should sense-check with @stevenleggdfe on whether this is the best way to do it.

Reply:
@misaka yes, we did it this way - because we didn't want to make it a field for all events, when many events won't have a table_name (say, page_visited events). We haven't had any issues extracting this from the STRUCT, although it does admittedly rely on a ~10-line user-defined function to do it each time. I don't think it matters that much either way, as long as the table_name is stored somewhere on each entity event.

Reply:
My thought was that you would be able to process the data with one less step if the table name was, say, a field called `data_table_name`. But happy to keep this this way for consistency.

Reply:
Given we're going to have many more `entity_*` events than we are `page_view`s, shall we just have a field in the schema?

Reply:
> Given we're going to have many more entity_* events than we are page_views, shall we just have a field in the schema?

That works for me. And agreed, we already have many request-specific fields, so having entity-specific fields that make things easier doesn't hurt.